pkg/io/opener.go

```diff
@@ -98,24 +99,30 @@ func (o opener) openGCS(path string) (*storage.ObjectHandle, error) {
 }
 
 // Reader will open the path for reading, returning an IsNotExist() error when missing
-func (o opener) Reader(ctx context.Context, path string) (io.ReadCloser, error) {
+func (o opener) Reader(ctx context.Context, path string, _ *blob.ReaderOptions) (io.ReadCloser, error) {
 	g, err := o.openGCS(path)
 	if err != nil {
 		return nil, fmt.Errorf("bad gcs path: %v", err)
 	}
 	if g == nil {
+		if strings.HasPrefix(path, "file://") {
+			path = strings.TrimPrefix(path, "file://")
+		}
 		return os.Open(path)
 	}
 	return g.NewReader(ctx)
 }
 
 // Writer returns a writer that overwrites the path.
-func (o opener) Writer(ctx context.Context, path string) (io.WriteCloser, error) {
+func (o opener) Writer(ctx context.Context, path string, _ *blob.WriterOptions) (io.WriteCloser, error) {
 	g, err := o.openGCS(path)
```

Comment:
> I'm not convinced we need/want io2

How do you want to implement the featuregate without adding a new package?

> Instead of replacing this whole package with io2, why don't we just continue extending this package to support S3? AKA:

People also asked about Azure and may ask about more providers in the future. What issue do you see with the abstraction used here?

> We can move the GCS related stuff into a gcs.go file and the s3 stuff into an s3.go or aws.go... Eventually these two files should go into their own package, and importing them can have an inti() which registers a particular schema like gs://.

Why should they go into their own package? That would mean that everything that uses storage needs to be aware of the underlying technology and that is IMHO exactly what we want to avoid and why we started to use the gocloud abstraction in the first place.

Reply:
> How do you want to implement the featuregate without adding a new package?

If we continue using the existing code path for `gs://` then we do not need a new package or featuregate (as the behavior isn't changing)

This PR should not need to make changes to tide aside from adding the flag for S3 and using it when constructing the opener. Instead we're changing all the calls to `Reader()` and `Writer()`

> What issue do you see with the abstraction used here?

Consumers of this interface should not have to import gcloud.dev/blob in order to use the Opener.

> Why should they go into their own package?

Not totally necessary, but it would provide an easy way for people who build their own images to reduce dependencies by changing what they import.

AKA if I wanted my prow distribution to not support honking, I can provide my own plugin-import package that drops this line: https://github.com/kubernetes/test-infra/blob/master/prow/hook/plugin-imports/plugin-imports.go#L36

I do not need/want to do that now, but we should keep it in mind.